cpp/ql/test/library-tests/dataflow/dataflow-tests/BarrierGuard.cpp

```diff
@@ -75,4 +75,13 @@ void bg_indirect_expr() {
   if (guarded(buf)) {
     sink(buf);
   }
-}
\ No newline at end of file
+}
+
+void test_guard_and_reassign() {
+  int x = source();
+
+  if(!guarded(x)) {
+    x = 0;
+  }
+  sink(x); // $ SPURIOUS: ast
+}
```

Comment:
How am I supposed to read test like these?

Reply:
`guarded` is a bad name for the barrier guard in this case. The function should really be read as `isSafe`. So this is a standard dataflow test, except that there's a function `guarded` (read: `isSafe`) to check if the value produced by `source()` is allowed to flow to `sink`.

So in this case, the value produced by `source` is either safe (i.e., `guarded(x)` is true), or the value is reassigned to 0.

And after these changes, IR dataflow realizes that the value of `x` is always either 0 or `guarded(x)` holds when reaching `sink(x)`.

Reply:
Thanks for the explanation. That helps!